### DISCUSSION TITLE
CRA-template prototype review

### DISCUSSION THREAD

**Discussion by azinit** [feature-sliced/cra-template#7] [General]
По задаче #5 , на момент https://github.com/feature-sliced/cra-template/commit/a3fb1ac928b67a8ee57ed229df4a01d6e497e627

---

Для начала - спасибо большое, что смог заняться этим! Нанесешь непоправимую и мучительную пользу сообществу!

**Comment by azinit** [comment]
## Начну с похвалы)

[`praise:`](https://conventionalcomments.org/) классно, что так быстро забутстрапил темплейт, да и еще и для base и для ts!

> Хотя щас возможно из-за этого придется в двух местах уточнять, но пока не суть 😄 

![image](https://user-images.githubusercontent.com/42924400/144754448-351cc8eb-4e4d-4eb4-9386-492989a5442c.png)

**Comment by azinit** [comment]
[`praise:`](https://conventionalcomments.org/) Здорово, что базово заполнил мета инфу!

![image](https://user-images.githubusercontent.com/42924400/144754812-c986f51c-0f04-4def-acc7-32c6badf6cb2.png)

**Comment by azinit** [comment]
[`praise:`](https://conventionalcomments.org/) Классно, что получилось сделать небольшой по размеру темплейт

![image](https://user-images.githubusercontent.com/42924400/144754849-92b5ce49-526e-4893-8421-6b951d231684.png)

**Comment by azinit** [comment]
[`praise:`](https://conventionalcomments.org/) Хорошо, что удалось перенять формат ридмишек, выглядит классно)

![image](https://user-images.githubusercontent.com/42924400/144754966-0ba208de-2b7d-457c-bb42-e3a16da1a533.png)

**Comment by azinit** [comment]
## Теперь ~~придирки~~ ToImprove))

[`nitpick(critical):`](https://conventionalcomments.org/) Не хватает в ридмихах упоминания Feature Sliced

Можно взять хотя бы как у эффектора

[`thoughts:`](https://conventionalcomments.org/) Плюс в целом бы как-то еще пометить, что не просто теймплейт Feature Sliced, но и структура (с ссылкой на сайт офк)


![image](https://user-images.githubusercontent.com/42924400/144760793-c797a322-f743-4775-a6e2-6c8f20cd978e.png)

**Comment by azinit** [comment]
[`nitpick(meta):`](https://conventionalcomments.org/) Придирки по мета-инфе)

1) В `keywords` кажется можно добавить как минимум `["feature-sliced",  "feature-based", "feature-slices"`
2) В `description` лучше все же с большой буквы написать (`Feature Sliced typescript template`)
   - Или даже `The offcial FeatureSliced template for Create React App.` как у эффектора
3) А `version` я бы пока поставил `0.0.1` т.к. это все прототип пока

> Аналогично также и с `base` темплейтом)

![image](https://user-images.githubusercontent.com/42924400/144760845-8644049e-1070-479b-a05b-2c2a42b54dbf.png)

**Reply by azinit** [reply]
[`nitpick(meta):`](https://conventionalcomments.org/) Ну и еще по мета инфе😄 

4) В `repository` и `bugs` все же указал инфу о репозитории, как у [эффектора](https://github.dev/effector/cra-template)

![image](https://user-images.githubusercontent.com/42924400/144760989-069efd23-b933-4770-bbae-506036ee58d4.png)

**Comment by azinit** [comment]
[`nitpick(fundamentals, public-api):`](https://conventionalcomments.org/) Вот это прям вразрез с методологией идет)

Я бы добавил файлы реэкспорты в каждый компонент, и поправил импорты в `shared/ui/index.js` соответственно


> В `base` и `typescript` соответственно 👀 

![image](https://user-images.githubusercontent.com/42924400/144761045-51726aff-3bd5-4a7a-bb65-bca3bf85864f.png)

**Reply by azinit** [reply]
[`thoughts:`](https://conventionalcomments.org/) Да и раз мы делаем реэкспорт на весь `shared/ui` неплохо было бы комментарием добавить приписку о тришейкинге с ссылкой 🤔 

https://feature-sliced.design/docs/concepts/public-api#%D0%B2%D0%BE%D0%B7%D0%BC%D0%BE%D0%B6%D0%BD%D1%8B%D0%B5-%D0%BF%D1%83%D1%82%D0%B8-%D1%80%D0%B5%D1%88%D0%B5%D0%BD%D0%B8%D1%8F

**Reply by azinit** [reply]
Тут тоже офк)

![image](https://user-images.githubusercontent.com/42924400/144761834-986380bc-fda9-4757-a147-03b7288cc21f.png)

**Comment by azinit** [comment]
[`nitpick(fs, structure):`](https://conventionalcomments.org/) Так нам бы или `kebab-case` или `PascalCase` в файловой структуре придерживаться)

Надо бы выбрать
> JFYI, кортима активно приветствует именно `kebab-case`

![image](https://user-images.githubusercontent.com/42924400/144761116-c9818f81-8bca-4ef0-b25e-9ba8404905c1.png)


![image](https://user-images.githubusercontent.com/42924400/144761130-3b5436bb-714a-49d5-9e08-c0e18c057ffe.png)

**Reply by tednaaa** [reply]
я думал делать через PascalCase компоненты, а через kebab-case обычные модули :)

**Reply by azinit** [reply]
@tednaaa Надо бы что-то одно)
Иначе путанно, и будет раздрай

**Reply by azinit** [reply]
@feature-sliced/core что думаете?

**Reply by sergeysova** [reply]
Я за kebab-case все модули. Пусть структура определяет компонент это или нет, а не кейс символов.
Плюс избежим сразу огромного количества проблем с git и регистронезависимыми фс

**Reply by azinit** [reply]
Согл
\+ кому оч надо, переведут на PascalCase новые модули
(потом в доке мб отдельно распишем проблематику эту)

**Comment by azinit** [comment]
[`nitpick(segments, non-blocking):`](https://conventionalcomments.org/) Кмк тут как раз случай, когда необязательно все группировать в `ui` и хватило бы обычного

```
📁 pages/counter-page
     ui.jsx
     styles.module.css
     index.js
```

![image](https://user-images.githubusercontent.com/42924400/144761247-f8ed3753-8e4d-4069-bd1a-ec2a8c56c1c3.png)

![image](https://user-images.githubusercontent.com/42924400/144761174-f86446bb-b9dd-4f85-8ba7-bdefa9602b81.png)

**Reply by azinit** [reply]
[`thoughts:`](https://conventionalcomments.org/) Хотя не исключено, что меня просто зацепило излишнее дробление файлов в формате

`pages/counter-page/ui/counter-page.jsx`, когда можно было одним файлом обойтись

Так что некрит особо

**Comment by azinit** [comment]
[`nitpick(shared, routes):`](https://conventionalcomments.org/) Задумка с роутами интересная вышла!

Но при всем уважении к этому, кажется в рамках темплейта так вовсе не обязательно упарываться)

[`issue(shared, routes):`](https://conventionalcomments.org/)
Тем более, что вродь пока ни в чате, ни в кортиме пока не закреплялся активно вариант, когда в шейреде бы роуты хранились
(которые как бы завязаны на проект и роутинг)

[`suggestion(shared, routes):`](https://conventionalcomments.org/) Поэтому я бы прямо в `pages/routes.js` роуты писал, как у [кардбокса](https://github.com/cardbox/frontend/blob/master/src/pages/routes.ts)
   - И сразу видно - что при обращении к роутам, придется лезть в уровень страниц - что [осуждается](https://feature-sliced.design/docs/concepts/issues/routes) по большей части

![image](https://user-images.githubusercontent.com/42924400/144761307-268f3706-54df-415e-9239-f36166eca7e0.png)

**Reply by azinit** [reply]
@feature-sliced/core что скажете? Или мб зря набрасываю? 🤔

**Reply by tednaaa** [reply]
можно прямо в pages, согласен, но что делать, если к примеру внутри entities/foo нужно использовать history.push(routeNames.foo) либо что то связанное с конкретным роутом?

**Reply by azinit** [reply]
[Не завязываться на конкретный роут](https://feature-sliced.design/docs/concepts/issues/routes#%D0%B5%D1%81%D0%BB%D0%B8-%D0%BF%D1%80%D0%BE%D0%B8%D0%B3%D0%BD%D0%BE%D1%80%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D1%82%D1%8C) :)

**Reply by azinit** [reply]
Никто же нам не мешает передать через пропсы/параметры нужный урл со страницы, а мы их обрабтаем как надо

И не будет размазывания ответственности

**Reply by tednaaa** [reply]
т.е роуты лучше держать внутри pages/router а не в shared/router ?

**Reply by azinit** [reply]
Именно

Глянь как у кардбокса)
Там тоже люди читерят тем, чтоб их юзать из слоев ниже, но зато явно сразу видно - что мы лезем в чужую зону ответственности
Лучше хотя бы явно чтоб это было

https://github.com/cardbox/frontend/blob/master/src/pages/routes.ts
https://github.com/cardbox/frontend/blob/master/src/pages/paths.ts

> Не обязательно так же сложно - хватит рили статического объекта с роутами

**Comment by azinit** [comment]
[`nitpick(app, public-api):`](https://conventionalcomments.org/) Не хватает файла реэкспорта)

Даже хоть используется это только в руте, надо соблюдать концепции))

Иначе в раздрай уйдет

![image](https://user-images.githubusercontent.com/42924400/144761422-39be3522-9178-45fe-a52e-44fb4306813e.png)

**Comment by azinit** [comment]
[`nitpick(app, providers):`](https://conventionalcomments.org/)

![image](https://user-images.githubusercontent.com/42924400/144761479-6df3ac9e-bb2c-459e-b7a7-fd5c4d230582.png)

1) Некрит, но генерацию роутинга вынес бы все же на уровень pages
    У нас ведь там роуты хранятся с маппингом - там же поидее и саму генерацию в jsx роутинг остается сделать (см. как [у кардбокса](https://github.com/cardbox/frontend/blob/master/src/pages/index.ts))
   Я бы понял, если бы и роуты мы на уровне app объявляли... Хотя тогда еще больше проблем было бы))
   Энивей, кажется знание про роутинг - это [дело pages](https://feature-sliced.design/docs/concepts/issues/routes). А app только подключает все нужное
   
2) А тут я бы как раз заморочился и выделил бы отдельный HOC-провайдер `withRouter`
    Чтобы люди не лепили с самого начала всю логику инициализации в одном жирном компоненте, а разбивали ее, как на скрине
    
    Можно посмотреть реализацию [тут](https://github.com/ani-team/github-client/blob/workshop/feature-sliced/src/app/providers/index.ts) или [тут](https://github.com/select-name/sharead-frontend/tree/master/src/app/hocs)
    > Называть все же лучше `providers`, т.к. у того же Vue нет HOC для этого, но есть концепция `plugins`, которые то же самое делают
    ![image](https://user-images.githubusercontent.com/42924400/144761607-60e27e6b-da10-43c7-85aa-84be5b725c0b.png)

**Comment by azinit** [comment]
[`thoughts(non-blocking, model, ui):`](https://conventionalcomments.org/) Возможно прозвучит как бред, но есть идея вынести отдельно в `pages/counter-page/model.js` хук типа `useCounter`, который бы давал все хендлеры нужные для "тупого" ui

[`thoughts:`](https://conventionalcomments.org/) Чтобы идея разделения модели и ui была сразу, и чтобы люди не гнушались туда хотя бы такую логику выносить
(если СТМ не завели, то хотя бы так разделять

![image](https://user-images.githubusercontent.com/42924400/144761700-722777a1-6d3a-4fd8-8b68-5663f95297b5.png)

**Reply by azinit** [reply]
И этот хук возвращал бы как раз полный объект в формате

`{ counter, setCounter, increment, decrement }`

И с юзколлбэками офк)

**Comment by azinit** [comment]
[`thoughts(non-blocking):`](https://conventionalcomments.org/) Опасно))

Но в целом некрит

![image](https://user-images.githubusercontent.com/42924400/144761804-7d8a845b-d2a6-49aa-b811-24a1f62baf98.png)

**Reply by tednaaa** [reply]
а как? :) отдельно каждый пропс?

**Reply by azinit** [reply]
Ну там же ты по-факту 2-3 параметра передаешь

У тебя нет цели полностью кнопку воспроизвести

А пропс дриллинг такой опасен малясь

**Comment by azinit** [comment]
[`suggestion(promote):`](https://conventionalcomments.org/) Надо продвигать слайсы в том числе и лого кмк)))

Кажется можно сделать прям как у эффектора и заменить основное реактовское лого на слайсовое
![image](https://user-images.githubusercontent.com/42924400/144762077-643c5bcb-6bd0-4ced-bb70-4f7753207814.png)

**Reply by azinit** [reply]
Вот тут файлы все нужные)

Поменять бы в обоих темплейтах

https://drive.google.com/drive/folders/15qy6E7OOkJiUWdY3I1DeB1L_3YsDoE8z?usp=sharing

**Comment by azinit** [comment]
[`praise:`](https://conventionalcomments.org/) Кста очень классно, что baseUrl сразу задал! Почему-то не заметил сначала)

![image](https://user-images.githubusercontent.com/42924400/144762192-bf277471-3c08-452b-94d9-87e042e76a0b.png)
![image](https://user-images.githubusercontent.com/42924400/144762199-f7bc76b8-14bb-4fb6-aecc-46acd328b688.png)

**Comment by azinit** [comment]
[`suggestion(meta):`](https://conventionalcomments.org/) Ну в манифесте, собсна как и [у эффектора](https://user-images.githubusercontent.com/42924400/144762222-581d744e-0870-4cfe-af6e-2c51e8b44b6d.png), сделал бы описание)

А то у нас пока оч дефолтный манифест

![image](https://user-images.githubusercontent.com/42924400/144762243-4c5327e3-3b6e-4afb-aaf9-8a1ea6ed2549.png)

**Comment by azinit** [comment]
[`question:`](https://conventionalcomments.org/) Я не тестил конечно, но точно ли там префикс `base` обязателен?
У эффектора вродь без него обошлось 🤔 

https://github.com/effector/cra-template/blob/5792e4be311cd9d97e45cb5336f15d316b71cae7/base/README.md#L18

![image](https://user-images.githubusercontent.com/42924400/144762263-5f1f7b76-5d00-4fab-9221-e98640158517.png)

**Reply by tednaaa** [reply]
оу, по поводу этого точно не подскажу :)

**Comment by azinit** [comment]
[`nitpick(critical):`](https://conventionalcomments.org/) Я не сторонник нанооптимизаций, но кмк прям отдельную зависимость в шаблон тянуть малясь излишне 🤔 

Если получится без нее реализовать нормально кнопку, было бы здорово
_(можно сделать костыльный дедовский кастомизатор класса, но приписать мол - можете заюзать `classnames`/`clsx` или аналоги)_

![image](https://user-images.githubusercontent.com/42924400/144762316-23c7bc73-4878-4cc0-a6ab-5597f3b662f1.png)

**Comment by azinit** [comment]
Фух,  ну вроде ~~до каждого файла докопался~~ все базово глянул)

@tednaaa спасиб тебе большое , что взялся - выглядит как оч мощный кандидат на прототип темплейта!

Думаю потом можно будет и полноценный забабахать, со всеми слоями 🚀

**Reply by azinit** [reply]
@tednaaa Энивей, тут не было цели набрасывать на твои старания

А офк - чуть подзапариться над концептом темплейта, чтоб не стыдно было публиковать, даже как `0.0.1`

> Тем более, не хочется, чтобы люди с самого начала вредные практики начали применять

**Reply by tednaaa** [reply]
ничего страшного, наоборот я рад что хотя бы тут получаю код ревью))

**Comment by azinit** [comment]
@feature-sliced/core Мб вам еще будет что добавить, пишите)

**Comment by azinit** [comment]
[`thoughts:`](https://conventionalcomments.org/) @tednaaa Кста кажется еще, что в ридмихах темплейтов было бы нелишним добавить, что все доп. слои (entities/features/widgets) можно смело добавлять - но эти темплейты максимально упрощенные

И тоже ссыль на статью со слоями закинуть 🤔

**Reply by tednaaa** [reply]
окей, тоже добавлю, но это наверно останется в след выходные 😄

**Reply by azinit** [reply]
Ага)

Либо можно сделать рили как у https://github.com/cristobalgvera/ez

Т.е. `src/entities/.gitkeep`
